### ISSUE TITLE
Object.php

### ISSUE THREAD

**Issue by Criptos** [DanielnetoDotCom/YouPHPTube#831]
I'm havng an issue. The object Object, in the function save, it decides if is an insert or update based on the $this->id. 
I haven an object initializaed without id, so save() would insert it, but it keeps sending 

`INSERT INTO events_tickets ( `id`,`ticketType`,`price`,`code`,`qrCode`,`clientName`,`clientEmail`,`dateAssigned`,`cancelled`,`validatedDate`,`creationDate`,`events_id` ) VALUES ( NULL ,  'Entrada Genral' ,  '0' ,  'PSOR83T18' ,  'iVBORw0KGgoAAAANSUhEUgAAAVwAAAFcAQMAAAC+2JRvAAAABlBMVEX///8AAABVwtN+AAAA6klEQVRoge3XQRLCIAyFYW7A/W/JDVAaCKEw6kYXvv9NF5n41VUmhZQIIYQQQn6S6imtTrkX3gTr4hIKk2VtgoVxn6IL905sgsGxAIPTvb+/CFbGsXiSvQkWxR5fL+WzUy743/E9+R0AK+H4ubHxsbfStmTAWnhccv2nareb40SBpXA4liwTNR6wLK5+UrWtMl5pg2TbBqyKzfSVkrs5zhtYDocYi/8AVsbVM1iNRxSwMi6zMDA3DFge+xdneeJ0geWxd1qOp1ywKq4jvl5eTR1YAM8iDxYuv2Bd7HEcziRgZUwIIYQQ8rU8AMXilx2v8Yq1AAAAAElFTkSuQmCC' ,  '' ,  '' ,  NULL ,  'false' ,  '0000-00-00 00:00:00' ,  '2018-07-05 14:18:08' ,  NULL )`

The id send it as null, and then makes an db error about id not being allowed to be null.. 

The soluton would be to ignore the id field when insert, and it shouldn't breake anything...Do I proceed?

**Comment by DanielnetoDotCom** [comment]
Try like I did here https://github.com/DanielnetoDotCom/YouPHPTube/blob/7f8a9be1285e1da133bc3e9dfa96b30b4c6cf9d3/objects/comment.php#L77

Overwrite the save method

and test the parameter, if it is empty place a string 'null', so your field will be null instead a empty string

In example

```
public function save() {
       if(empty($this->your_field)){
            $this->your_field = 'NULL';
        }
        return parent::save();
}

```

**Comment by Criptos** [comment]
I tried that, but database keep telling that id field cannot be null...   
Id is declared as auto incrementable unique index not null... 
I did this to fix it:

`            $sql = "INSERT INTO " . static::getTableName() . " ( ";
            if (($key = array_search("id", $fieldsName)) !== false) {
                unset($fieldsName[$key]);
            }            
`

**Comment by DanielnetoDotCom** [comment]
hi, do you mean the primary key?
is it auto increment?
can you please share the code you are using?

**Comment by Criptos** [comment]
TAble;
`| events_tickets | CREATE TABLE `events_tickets` (
  `id` int(22) NOT NULL AUTO_INCREMENT,
  `ticketType` varchar(50) NOT NULL,
  `price` decimal(10,2) NOT NULL DEFAULT '0.00',
  `code` varchar(12) DEFAULT NULL,
  `qrCode` blob NOT NULL,
  `clientName` varchar(200) NOT NULL,
  `clientEmail` varchar(200) NOT NULL,
  `dateAssigned` datetime DEFAULT '0000-00-00 00:00:00',
  `cancelled` enum('false','true') DEFAULT 'false',
  `validatedDate` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `creationDate` datetime NOT NULL,
  `events_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  KEY `events_id` (`events_id`),
  KEY `ticketType` (`ticketType`),
  KEY `code` (`code`),
  CONSTRAINT `events_tickets_fk0` FOREIGN KEY (`events_id`) REFERENCES `events` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=22 DEFAULT CHARSET=latin1 `

The object (snip) declaration: 
`class Ticket extends ObjectYPT {

    /**
    * @id id of the eventName
    */
    protected $id; 
`

The save function: 
`    function save()
    {
    
        if(empty($this->id))
        {   
            $this->cancelled='false'; 
            $this->creationDate=date("Y-m-d H:i:s");
            $this->price=0.00;
            $this->clientName="";
            $this->clientEmail="";
            $this->validatedDate="0000-00-00 00:00:00"; 
        }
        return parent::save();
    }
`

**Comment by DanielnetoDotCom** [comment]
And if you make this ...

```
 function save(){
    if(empty($this->id))    {   
        $this->id = 0;
        $this->cancelled='false'; 
        $this->creationDate=date("Y-m-d H:i:s");
        $this->price=0.00;
        $this->clientName="";
        $this->clientEmail="";
        $this->validatedDate="0000-00-00 00:00:00"; 
    }
    return parent::save();
}
```

Usually when I instantiate an object to insert it i do ...

` $ticket = new Ticket(0);`

And I have no problem with this

**Comment by DanielnetoDotCom** [comment]
Also a tip, if you add a field named 

created and modified on the database, they will be automatic filled when you insert and update;

**Comment by Criptos** [comment]
aaaa... sweet :) 

thanks.

**Comment by DanielnetoDotCom** [comment]
Does it works?

**Comment by Criptos** [comment]
Yes it worked with the 0... 
 
ANother question, I had a weird situation where the data weren't saving.. I mean. 
I loaded the object.
I changed 1 data of the object. 
I saved the object. 
I create a new instance of the object in anther method. 
THe object is with out the updates. 
If I went to the database, the updates where made. 
 
To solve this I passed the modified object as a reference... instead of te instantiate the object... 
 
Any ideas?
